elasticsearch/backoff.go

```diff
@@ -0,0 +1,56 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elasticsearch
+
+import (
+	"math"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/outputs/elasticsearch"
+)
+
+type backoffFunc func(int) time.Duration
+
+var (
+	// DefaultBackoffConfig is the default backoff configuration used for
+	// es clients.
+	DefaultBackoffConfig = elasticsearch.Backoff{
+		Init: time.Second,
+		Max:  time.Minute,
+	}
+
+	// DefaultBackoff is an exponential backoff configured with default
+	// backoff settings.
+	DefaultBackoff = exponentialBackoff(DefaultBackoffConfig)
+)
+
+func exponentialBackoff(b elasticsearch.Backoff) backoffFunc {
+	// Convert max from nanoseconds to seconds
+	maxSeconds := float64(b.Max / time.Second)
+	return func(attempts int) time.Duration {
+		// Attempts starts at 1, after there's already been a failure.
+		// https://github.com/elastic/go-elasticsearch/blob/de2391/estransport/estransport.go#L339
+		if attempts == 1 {
+			return b.Init
+		}
+		next := math.Pow(float64(attempts), 2)
+		// Make sure we don't exceed max
+		next = math.Min(next, maxSeconds)
+		return time.Duration(next) * time.Second
```

Comment:
```suggestion
		next := b.Init * (1<<(attempts-1))
		if next > b.Max {
			next = b.Max
		}
		return next
```

We need to take `b.Init` whether or not attempts is 1, and it should be 2^N not N^2 (sorry if you did it this way based on my message in Slack, I wrote that backwards).

Reply:
Ah, ok. I was thinking about `backoff.init` and wasn't sure if it meant "wait an initial X seconds, and then start an exponential backoff", or "wait X seconds, and then each additional backoff is exponential based off the initial backoff". I read the docs in `apm-server.yml` and it wasn't super clear:

```
After waiting backoff.init seconds, apm-server
tries to reconnect. If the attempt fails, the backoff timer is increased
exponentially up to backoff.max.
```

I'm not sure which way is "better", but wanting to wait an initial 20s, for example, and then starting the backoff from there (t0 = 20s, t1 = 21s, t2 = 22s, t3 = 24s, etc), is something that I could see making sense.

Reply:
I have seen implementations like that before, where there's an initial delay -- it's not how the [libbeat implementation](https://github.com/elastic/beats/blob/2d88ed0286494c9423011b02784c14e79c3f7b8a/libbeat/common/backoff/equal_jitter.go#L54-L73) works though. There the init value is used to initialise the backoff duration on first failure, and then that duration is exponentially increased up to the max.

I suggest we align with libbeat, excluding jitter for now at least, and consider clarifying the docs (which are copied from beats). I'd be surprised if anyone looks at those docs though tbqh; I expect (and hope) folks mostly rely on the defaults.

Reply:
sounds good :ok_hand: